### DISCUSSION TITLE
Why doesn't this work?

### DISCUSSION THREAD

**Discussion by robknight** [LegendApp/legend-state#105] [Q&A] [answered]
https://codesandbox.io/s/quirky-water-srz01o?file=%2Fsrc%2FApp.tsx

There are nodes, which have IDs, names, and children. I want to be able to show a node, recursively show its children, and add and remove nodes. I have an object which acts as a dictionary of id -> node, and each node stores the IDs of its children. The node dictionary is stored in an observable.

At the beginning, the root node is the only one that exists. There are two buttons: one which adds a child to the root node, and another which adds a child to the root node and changes the `name` property on the root node.

When I click the first button, the new node is added to the state, but no DOM update occurs. When I click the second button, the DOM update occurs - presumably because the `name` property has changed, and this triggers a re-render.

How can I trigger a re-render when the `children` property changes?

**Comment by jmeistrich** [comment, marked as answer]
It's because of this:

```
state.nodes.root.children.set((children) => {
  children.push(id);
  return children;
});
```

Because you are returning the same object as the current value, it thinks nothing has changed so it doesn't run the notifications. You're doing the push on the raw underlying data which doesn't know anything about observables, and then the set method only sees that it is the same array object. So you can either:
1. Return a new array object, like `[...children, id]`
2. Use the `.push(...)` method on the observable array to just push it directly: `state.nodes.root.children.push(id)`
